<a href="https://colab.research.google.com/github/a01736552-design/Valores_Nulos_Act/blob/main/Act_2_1_Valores_Nulos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
#Importamos las librerias pandas, numpy y matplotlib respectivamente
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

**VALORES NULOS**

In [4]:
#Carga desde un archivo .csv sin indice
data = pd.read_csv('Citas_Digital_Filtrado.csv')
#Los nombres de columna traen espacios extra (ej. 'BDC ', 'Nombre Cliente '), los limpiamos
data.columns = data.columns.str.strip()
data.head(5)

,BDC,Fecha,Nombre Cliente,Estatus de Lead,Potencial de compra,PDM,SDC,Venta,Asesor Asignado,¿Por que no ha visitado la agencia?,¿Por que no hay solicitud de credito?,¿Por que no hubo prueba de manejo ?,¿Por que no hubo proceso wow?,¿Por que se asigno antes de vistar piso?
0,Saul,NaN,Guillermo,Finalizado,0.9,1.0,0.0,0.0,Nancy,"Ya visito la Agencia el dia jueves 17, intere...","Pide arrendamiento, unicamente se esta en espe...","Si, hizo prueba de manejo","Si recibio procesos wow, estuvo en la agencia ...","No, se espero a que visitara la agencia para s..."
1,Saul,NaN,Roberto,Finalizado,0.4,0.0,0.0,0.0,Victor,Ya visito la Agencia el dia 3 de abril,Por que primero queria manejar las unidad,No habia unidad GS8 HEV para prueba de manejo,"Si recibio procesos wow, sin embargo no quizo ...",Se asigno hasta que visitara Piso
2,Saul,NaN,Irving,Finalizado,0.85,0.0,0.0,0.0,Alan,Ya visito la agencia el dia 21 de abril,"Firmo solicitud, sin embargo despues comento q...",Por que no habia unidad para prueba de manejo,"Si se realizo el proceso wow, los clientes se ...",Se asigno hasta que visitara Piso
3,Saul,NaN,Oswaldo,Venta,0.9,0.0,0.0,1.0,Mauricio,El cliente vive en Tabasco Villahermosa,Sera compra de contado,"Ya la manejo en Villahermosa, sin embargo por ...","Las veces que se le ha marcado, agradece mucho...","Necesitabamos sacar el apartado, por eso se as..."
4,Saul,NaN,Rodrigo,Finalizado,0.9,1.0,0.0,0.0,Mauricio,"Visito la agencia el mes pasado, y nos visita ...",Hoy visita nuevamente la agencia y se firma so...,"La primera vez que nos visito, realizo PDM",Si recibio el proceso wow la primera vez que n...,Se asigno hasta que visitara Piso la primera ...


In [5]:
#Revisamos tipos de dato y dimensiones del dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 673 entries, 0 to 672
Data columns (total 14 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   BDC                                       657 non-null    object 
 1   Fecha                                     72 non-null     object 
 2   Nombre Cliente                            652 non-null    object 
 3   Estatus de Lead                           638 non-null    object 
 4   Potencial de compra                       599 non-null    object 
 5   PDM                                       672 non-null    float64
 6   SDC                                       671 non-null    float64
 7   Venta                                     668 non-null    float64
 8   Asesor Asignado                           641 non-null    object 
 9   ¿Por que no ha visitado la agencia?       493 non-null    object 
 10  ¿Por que no hay solicitud de credito? 

In [6]:
#Identificar valores nulos por columna
valores_nulos = data.isnull().sum()
valores_nulos

,0
BDC,16
Fecha,601
Nombre Cliente,21
Estatus de Lead,35
Potencial de compra,74
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180


In [7]:
#Identificar valores nulos por dataframe (total)
valores_nulos = data.isnull().sum().sum()
valores_nulos

np.int64(2208)

In [8]:
#Realizamos una copia del dataframe
data3 = data.copy()

Fecha

Con 89% de valores nulos (solo 72 de 673 registros tienen una fecha real), imputar esta columna con la media, la moda o un texto fijo no aportaría información real, ya que casi todo su contenido sería inventado o repetido. Siguiendo la misma lógica de eliminación de columnas, la mejor decisión aquí es descartar la columna completa en vez de simular un dato que no existe.

In [9]:
#Primer método: eliminación de columna innecesaria
#axis = 1 para columnas (misma lógica que en el notebook original)
data3 = data3.drop("Fecha", axis=1)

In [10]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,16
Nombre Cliente,21
Estatus de Lead,35
Potencial de compra,74
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


BDC

Se usa ffill, asumiendo que registros consecutivos en el archivo suelen pertenecer al mismo BDC, por lo que replicar el valor previo es razonable.

In [11]:
#Segundo método de sustitución de valores nulos
#Sustituir valores nulos por valores no nulos hacia adelante "forward fill" ("ffill")
data3["BDC"] = data3["BDC"].ffill()

In [12]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,21
Estatus de Lead,35
Potencial de compra,74
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


Nombre Cliente

Aquí se usa sustitución por un valor de texto específico, ya que no existe un "nombre promedio" que tenga sentido imputar.

In [13]:
#Tercer método de sustitución de valores nulos
#Sustituir valores nulos por un string en concreto
data3["Nombre Cliente"] = data3["Nombre Cliente"].fillna("Cliente sin nombre registrado")

In [14]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,35
Potencial de compra,74
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


Estatus de Lead

Se usa relleno hacia atrás (`bfill`), tomando el estatus del registro inmediato siguiente como referencia de continuidad del flujo de leads.

In [15]:
#Cuarto método de sustitución de valores nulos
#Sustituir valores nulos por valores no nulos hacia atrás "backward fill" ("bfill")
data3["Estatus de Lead"] = data3["Estatus de Lead"].bfill()

In [16]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,74
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


Potencial de compra, probabilidad 0–1

La columna llegó como texto porque incluye capturas erróneas. Primero se convierte a numérico (los valores no numéricos quedan como nulo) y después se aplica sustitución con la mediana, más robusta que la media ante esos valores atípicos de captura.

In [17]:
#Quinto método de sustitución de valores nulos
#Convertimos a numérico (valores mal capturados se vuelven NaN) y sustituimos con la mediana
data3["Potencial de compra"] = pd.to_numeric(data3["Potencial de compra"], errors="coerce")
data3["Potencial de compra"] = data3["Potencial de compra"].fillna(round(data3["Potencial de compra"].median(), 2))

In [18]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


PDM — numérica

Solo 1 valor nulo y sin outliers extremos, por lo que se usa sustitución con el promedio (media).

In [19]:
#Sexto método de sustitución de valores nulos
#Sustituir valores nulos con promedio o media
data3["PDM"] = data3["PDM"].fillna(round(data["PDM"].mean(), 2))

In [20]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,0
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


SDC

Se usa sustitución con la mediana, priorizando robustez sobre la media.

In [21]:
#Séptimo método de sustitución de valores nulos
#Sustituir valores nulos con mediana
data3["SDC"] = data3["SDC"].fillna(round(data["SDC"].median(), 2))

In [22]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,0
SDC,0
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


Venta

Se usa sustitución por un valor numérico específico (0) y si no quedó registrada, se asume que no hubo venta.

In [23]:
#Octavo método de sustitución de valores nulos
#Sustituir valores nulos por un valor numérico en concreto
data3["Venta"] = data3["Venta"].fillna(0)

In [24]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,0
SDC,0
Venta,0
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


Asesor Asignado — variable categórica

Se usa relleno hacia adelante (`ffill`), igual razonamiento que en BDC, la continuidad entre registros consecutivos.

In [25]:
#Noveno método de sustitución de valores nulos
#Sustituir valores nulos por valores no nulos hacia adelante "forward fill" ("ffill")
data3["Asesor Asignado"] = data3["Asesor Asignado"].ffill()

In [26]:
#Corroboramos valores nulos
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,0
SDC,0
Venta,0
Asesor Asignado,0
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


Columnas de texto libre condicional

Estas 5 columnas son comentarios abiertos que solo aplican cuando la respuesta a la pregunta implícita es relevante para ese lead (por ejemplo "¿por qué no ha visitado la agencia?" solo tiene sentido si no la ha visitado). Un nulo aquí probablemente significa que la pregunta no aplicaba al caso, no que falte el dato. Se usa sustitución por un valor de texto específico (No aplica) en las cinco.

In [27]:
#Décimo método de sustitución de valores nulos
#Sustituir valores nulos por un string en concreto, en las columnas de texto condicional
columnas_condicionales = [
    "¿Por que no ha visitado la agencia?",
    "¿Por que no hay solicitud de credito?",
    "¿Por que no hubo prueba de manejo ?",
    "¿Por que no hubo proceso wow?",
    "¿Por que se asigno antes de vistar piso?"
]
for columna in columnas_condicionales:
    data3[columna] = data3[columna].fillna("No aplica")

In [28]:
#Corroboramos valores nulos en todo el dataframe
valores_nulos = data3.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,0
SDC,0
Venta,0
Asesor Asignado,0
¿Por que no ha visitado la agencia?,0
¿Por que no hay solicitud de credito?,0


In [29]:
#Corroboramos valores nulos en todo el dataframe (total)
valores_nulos = data3.isnull().sum().sum()
valores_nulos

np.int64(0)

Método de Eliminación de valores nulos

Como alternativa de referencia (no se usa para el archivo final), podemos ver lo que pasaría si en vez de sustituir se eliminaran directamente todas las filas con al menos un valor nulo. Se parte de data sin la columna Fecha (ya descartada por columna) para que la comparación sea justa frente a data3.

In [30]:
#Realizamos una copia del dataframe, ya sin la columna Fecha
data4 = data.drop("Fecha", axis=1)
#Identificar valores nulos por columna antes de eliminar filas
valores_nulos = data4.isnull().sum()
valores_nulos

,0
BDC,16
Nombre Cliente,21
Estatus de Lead,35
Potencial de compra,74
PDM,1
SDC,2
Venta,5
Asesor Asignado,32
¿Por que no ha visitado la agencia?,180
¿Por que no hay solicitud de credito?,294


In [31]:
#Eliminamos filas con nulos
#No se agrega nada para eliminar filas y 1 para eliminar columnas
data4 = data4.dropna()
data4

,BDC,Nombre Cliente,Estatus de Lead,Potencial de compra,PDM,SDC,Venta,Asesor Asignado,¿Por que no ha visitado la agencia?,¿Por que no hay solicitud de credito?,¿Por que no hubo prueba de manejo ?,¿Por que no hubo proceso wow?,¿Por que se asigno antes de vistar piso?
0,Saul,Guillermo,Finalizado,0.9,1.0,0.0,0.0,Nancy,"Ya visito la Agencia el dia jueves 17, intere...","Pide arrendamiento, unicamente se esta en espe...","Si, hizo prueba de manejo","Si recibio procesos wow, estuvo en la agencia ...","No, se espero a que visitara la agencia para s..."
1,Saul,Roberto,Finalizado,0.4,0.0,0.0,0.0,Victor,Ya visito la Agencia el dia 3 de abril,Por que primero queria manejar las unidad,No habia unidad GS8 HEV para prueba de manejo,"Si recibio procesos wow, sin embargo no quizo ...",Se asigno hasta que visitara Piso
2,Saul,Irving,Finalizado,0.85,0.0,0.0,0.0,Alan,Ya visito la agencia el dia 21 de abril,"Firmo solicitud, sin embargo despues comento q...",Por que no habia unidad para prueba de manejo,"Si se realizo el proceso wow, los clientes se ...",Se asigno hasta que visitara Piso
3,Saul,Oswaldo,Venta,0.9,0.0,0.0,1.0,Mauricio,El cliente vive en Tabasco Villahermosa,Sera compra de contado,"Ya la manejo en Villahermosa, sin embargo por ...","Las veces que se le ha marcado, agradece mucho...","Necesitabamos sacar el apartado, por eso se as..."
4,Saul,Rodrigo,Finalizado,0.9,1.0,0.0,0.0,Mauricio,"Visito la agencia el mes pasado, y nos visita ...",Hoy visita nuevamente la agencia y se firma so...,"La primera vez que nos visito, realizo PDM",Si recibio el proceso wow la primera vez que n...,Se asigno hasta que visitara Piso la primera ...
...,...,...,...,...,...,...,...,...,...,...,...,...,...
479,Antonio,Carlos León,Activo,0.7,0.0,0.0,0.0,Pedro Gonzalez,Ya visito la agencia,Ya esta en proceso,Ya realizo prueba de manejo,Recibio la atencion,Se asigno en la visita
483,Antonio,Christofer,Activo,0.65,0.0,0.0,0.0,Mauricio,Ya visito la agencia,Ya esta en proceso,Ya realizo prueba de manejo,Recibio la atencion,Se asigno en la visita
488,Antonio,MARÍA DEL CARMEN GRADOS,Activo,0.6,0.0,0.0,0.0,Aurelio,Ya visito la agencia,Ya esta en proceso,Ya realizo prueba de manejo,Recibio la atencion,Se asigno en la visita
490,Antonio,Juan José castellon,Activo,0.8,0.0,0.0,0.0,Ivan,Ya visito la agencia,Ya esta en proceso,Ya realizo prueba de manejo,Recibio la atencion,Se asigno en la visita


In [32]:
#Identificar valores nulos por columna
valores_nulos = data4.isnull().sum()
valores_nulos

,0
BDC,0
Nombre Cliente,0
Estatus de Lead,0
Potencial de compra,0
PDM,0
SDC,0
Venta,0
Asesor Asignado,0
¿Por que no ha visitado la agencia?,0
¿Por que no hay solicitud de credito?,0


**Comparamos DataFrame3 (Sustitución) y DataFrame4 (Eliminación)**

In [33]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 673 entries, 0 to 672
Data columns (total 13 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   BDC                                       673 non-null    object 
 1   Nombre Cliente                            673 non-null    object 
 2   Estatus de Lead                           673 non-null    object 
 3   Potencial de compra                       673 non-null    float64
 4   PDM                                       673 non-null    float64
 5   SDC                                       673 non-null    float64
 6   Venta                                     673 non-null    float64
 7   Asesor Asignado                           673 non-null    object 
 8   ¿Por que no ha visitado la agencia?       673 non-null    object 
 9   ¿Por que no hay solicitud de credito?     673 non-null    object 
 10  ¿Por que no hubo prueba de manejo ?   

In [34]:
data4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 318 entries, 0 to 491
Data columns (total 13 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   BDC                                       318 non-null    object 
 1   Nombre Cliente                            318 non-null    object 
 2   Estatus de Lead                           318 non-null    object 
 3   Potencial de compra                       318 non-null    object 
 4   PDM                                       318 non-null    float64
 5   SDC                                       318 non-null    float64
 6   Venta                                     318 non-null    float64
 7   Asesor Asignado                           318 non-null    object 
 8   ¿Por que no ha visitado la agencia?       318 non-null    object 
 9   ¿Por que no hay solicitud de credito?     318 non-null    object 
 10  ¿Por que no hubo prueba de manejo ?       3

El resultado de dropna  queda con 0 filas. Con 13 columnas y varias de ellas con cientos de nulos (hasta 320 en una sola columna), prácticamente todas las filas tienen al menos un valor nulo en alguna columna, así que basta uno solo para descartar la fila completa. Esto confirma que la eliminación no es viable para este archivo  (perdería el 100% de los registros) y refuerza que la combinación de eliminar la columna Fecha + sustituir el resto por columna data3 es lo mejor, ya que se conservan las 673 filas originales con información real en cada campo.

DataSet sin valores nulos a CSV

In [35]:
#Convertir DataFrame a CSV (sin la columna de índice)
data3.to_csv("Citas_Digital_Limpio.csv", index=False)

In [36]:
#Descargar archivo limpio en csv (solo funciona en Google Colab)
try:
    from google.colab import files
    files.download("Citas_Digital_Limpio.csv")
except ImportError:
    print("No estamos en Google Colab: el archivo 'Citas_Digital_Limpio.csv' ya quedó guardado en la carpeta de trabajo.")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>